### ISSUE TITLE
feature requests and future roadmap

### ISSUE THREAD

**Issue by kirk86** [pytorch/pytorch#6164]
PyTorch Feature Requests and Future RoadMap
--------------------------------
Hello everyone and thank your for this great project. I would like to point to some things that I find important for the future growth of pytorch. 

1. Easy name conventions throughout the library. I believe that some things should be renamed so that it resonates and are easier to remember. This will also lower the entry bar for new people exposed to pytorch.

Example:
```
import torch
import torchvision
import torchvision.models as models
model = models.densenet201()

to iterate over the layers I'll have to do the following

for idx, m in enumerate(model.named_modules):
```

The `named_modules` or `named_children` doesn't mean anything to me and as soon as I use them my brain has already forgotten about it. A more sane naming good be `model.layers` that seems to be a universal convention pretty much since other major libraries are using it.

I would also love to see something like `model.layers.add(x)`,  `model.layers.pop(x)`, `model.layers.get(x)`, `model.save()`, `model.load()`, etc.

2. Model dissection. 
More often we find ourselves in a scenario where we would like to get the output of a complex model at different levels.

Example:
```
let's say we have the model = models.densenet201()
to get the output at different levels we would have to do something like this
model1 = nn.Sequential(*list(model.children())[:4])
model2 = nn.Sequential(*list(model.children())[:7])

Which feels more of a hack to me TBH.

I would much  rather prefer to do the following:

model1 = model.layers.layer_name or model.layers[4].output
```

3. Dataset preprocessing and transformation pipeline switch between cpu/gpu
Many times we find ourselves the need to do the data preprocessing on the gpu especially in the scenario where there is an abundance. But the choice at the moment is inexistent. There should be a flag for the user to easily switch between cpu and gpu upon request in order to chose where the data preprocessing should take place.

4. Addition of negative indexing in Tensors.

5. Integration and applicability with already pre-existing tools.
Some tools that already pre-exist and are nice to have an out of the box support and integration with them. For instance tensorboard is great but at the moment we need to rely on third party developed plugins for this to work. Instead it would be better if pytorch had the ability to dump logs that can be consumed by tensorboard directly. 

5. Serving models is completely inexistent at the moment of writing.

For anyone else please feel free to add anything else you might think is important for future directions.
Basically I just watched the whole tf dev summit and the one advantage that pytorch might had against tf was the dynamic models but now that's gonna with eager mode. And since folks at tf have faster release cycles they can fail and iterate faster on concepts and ideas. A lot of people where nagging about the language choices which where horrific but since then they have iterated over a lot and brought it up to a sane level `tf.layers`.

These things make me wonder about the future directions and where pytorch is heading?

Thank you!

**Comment by Stonesjtu** [comment]
1. `torch.save` and `torch.load` may work as `model.save` and `model.load`.
2. I don't think `pytorch` uses `Layer`
3. `torch.set_default_tensor_type('torch.cuda.DoubleTensor')` should probably solve your problem.
4. `tensor[-4]` is already supported.
5.  refer to https://github.com/lanpa/tensorboard-pytorch
6.  agree with you.

Correct me if I misunderstand.

**Comment by kirk86** [comment]
@Stonesjtu Thank you for the input.

> torch.set_default_tensor_type('torch.cuda.DoubleTensor') should probably solve your problem.

Does that mean that whenever I construct the Dataset api all the preprocessing is gonna happen on the gpu? My understanding from reading around the forums was that something like that didn't exist unless I'm mistaken?

**Comment by Stonesjtu** [comment]
Technically that mean when you runs `b = torch.Tensor(5,5)`, `b` will be on GPU by default. I don't quite understand the need for pre-processing data on GPU actually.

**Comment by kirk86** [comment]
@Stonesjtu 

> I don't quite understand the need for pre-processing data on GPU actually.

Say you have many gpus, instead of doing the pre-processing on cpu you could leverage the gpus that are sitting there doing nothing.  That would speed up things a lot and during training the actual gpus that are responsible for the computation won't have to wait the cpu to finish to load data into gpu memory.  Things can happen in parallel especially if your pre-processing is computationally expensive.

**Comment by SsnL** [comment]
You can already to data reprocessing on GPU in dataloader workers with spawn method.

**Comment by kirk86** [comment]
@SsnL can you provide some pointers or minimal working example?

**Comment by goldsborough** [comment]
`nn.Module` represents a tree of modules. It has nothing to do with neural network layers.

**Comment by kirk86** [comment]
@goldsborough Thanks Peter, but I was thinking how hard it is to rename things. Just to have a consistency across libraries. One can transfer previous knowledge acquired by using some other library directly to pytorch without overhead. That would make the transition  easier for newcomers and increase the visibility of the library.

**Comment by kirk86** [comment]
@apaszke Thanks for the reply. I wasn't of course implying to have a single library 😆 . I understand that there are differences but sane easy to remember naming conventions that register immediately to your long term memory are quite helpful to get to speed up with the library. But I do get it, I understand the reason. Thanks!
